src/coreclr/jit/codegenxarch.cpp

```diff
@@ -2752,26 +2752,42 @@ void CodeGen::genCodeForInitBlkUnroll(GenTreeBlk* node)
         src = src->AsUnOp()->gtGetOp1();
     }
 
+    unsigned size = node->GetLayout()->GetSize();
+
+    // An SSE mov that stores data larger than 8 bytes may be implemented using
+    // multiple memory accesses. Therefore, the JIT must not use such stores
+    // when INITBLK zeroes a struct that contains GC pointers and can be visible
+    // to other threads (i.e. the struct is not located on the stack).
+    // Therefore, on x64 the JIT will not use SIMD stores for such structs and
+    // will allocate a GP register for the node.
+    // On x86 the JIT will use 8 bytes stores (movq) for structs that are
+    // larger than 16 bytes.
+    const bool canUse16BytesSimdMov = !node->MustUsePointerSizeAtomicStores();
+
     if (!src->isContained())
     {
         srcIntReg = genConsumeReg(src);
     }
     else
     {
-        // If src is contained then it must be 0 and the size must be a multiple
-        // of XMM_REGSIZE_BYTES so initialization can use only SSE2 instructions.
+        // If src is contained then it must be 0.
         assert(src->IsIntegralConst(0));
-        assert((node->GetLayout()->GetSize() % XMM_REGSIZE_BYTES) == 0);
+
+#ifdef TARGET_AMD64
+        assert(canUse16BytesSimdMov);
+        assert(size % 16 == 0);
+#else
+        assert(size >= XMM_REGSIZE_BYTES);
+        assert(size % 8 == 0);
+#endif
     }
 
     emitter* emit = GetEmitter();
-    unsigned size = node->GetLayout()->GetSize();
 
     assert(size <= INT32_MAX);
     assert(dstOffset < (INT32_MAX - static_cast<int>(size)));
 
-    // Fill as much as possible using SSE2 stores.
-    if (size >= XMM_REGSIZE_BYTES)
+    if (AMD64_ONLY(canUse16BytesSimdMov &&)(size >= XMM_REGSIZE_BYTES))
```

Comment:
Why is this check for x64 only, shouldn't the same restriction exist for x86?

Reply:
For x86 we will still emit `movq [m64], xmm` for structs larger than 16 bytes - since one such instruction will save us from emitting `mov [m32], r32` `mov [m32+4], r32`. So it really not important whether we can use `movdqu` or not on x86.

Reply:
The checks here, or maybe the names, seem a bit confusing then.

I'd expect this to be something more like `bool canUseSimdMov = !node->MustUsePointerSizeAtomicStores() && (size >= minSimdCopySize)`. `minSimdCopySize` would then be `16` on x64 and `8` on x86.

It would then also require that `(size % minSimdCopySize) == 0`

Reply:
I guess the logic I stated isn't quite exact.

The general point I'm trying to make is that this logic isn't exactly straightforward to follow with the logic for `movdqu` and `movq` and `x64` and `x86` all being together.

Reply:
Yes, the size of a struct doesn't matter - we will not use SIMD operations for zeroing on either amd64 or x86.
I think I could do this to make the logic clearer
```c++
#ifdef TARGET_AMD64
// On Amd64 if the JIT needs to guarantee atomicity for stores in INITBLK
// the codegen will avoid using SIMD stores and instead will use mov [m64], r64 for zeroing.
const bool willUseSimdMov = canUse16BytesSimdMov && (size >= XMM_REGSIZE_BYTES);
#else 
// On x86 even we can't use movdqu for zeroing it's still more beneficial to use
// movq [m64],xmm for structs larger than 16 bytes.
const bool willUseSimdMov = (size >= 16);
#endif
if (willUseSimdMov)
{

}
```

@tannergooding Would this help?

Reply:
Yes. I think that helps clarify both the readability and the intent for the different paths
-- It's not quite clear why we do `(size >= 16)` rather than `(size >= 8)` for x86, but I think that's just existing behavior

Reply:
Also, it might be worth renaming `canUse16BytesSimdMov` to just be `isReferenceOrContainsReferences` or something similar. That's the actual term on the managed side (https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.runtimehelpers.isreferenceorcontainsreferences?view=net-5.0) and is the limiting factor behind `movdqu` vs `mov`

Reply:
Yes, the boundary is to be somewhat consistent with current behavior.

Reply:
I would suggest the following name - `isOnHeapAndContainsReferences` or `maybeBeOnHeapAndContainsReferences` 
since it is definitely not about the type being a reference type and it is not *only* about a value type that contains references. 

It's about  a value type that contains references **and** not guaranteed to be allocated on the stack.

Reply:
@tannergooding I pushed the change renaming the method name and updating the comment. Does it read better now?

Reply:
Yes, its much easier to follow now. Thanks so much!